src/service/visibility-impl.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getService} from '../service';
+import {resourcesFor} from '../resources';
+import {viewportFor} from '../viewport';
+
+/**
+ * This type signifies a callback that gets called when visibility conditions
+ * are met.
+ * @typedef {function()}
+ */
+let VisibilityListenerCallbackDef;
+
+/**
+ * @typedef {Object.<string, JSONObject|VisibilityListenerCallbackDef>}
+ */
+let VisibilityListenerDef;
+
+/**
+ * Allows tracking of AMP elements in the viewport.
+ *
+ * This class allows a caller to specify conditions to evaluate when an element
+ * is in viewport and for how long. If the conditions are satisfied, a provided
+ * callback is called.
+ */
+export class Visibility {
+
+  /** @param {!Window} */
+  constructor(win) {
+    this.win_ = win;
+
+    /**
+     * key: resource id.
+     * value: [{ config: <config>, callback: <callback>}]
+     * @type {Object.<string, Array.<VisibilityListenerDef>>}
+     * @private
+     */
+    this.listeners_ = {};
+
+    /** @private {Array.<!Resource>} */
+    this.resources_ = [];
+
+    /** @private */
+    this.boundScrollListener_ = null;
+  }
+
+  /**
+   * @param {!JSONObject} config
+   * @param {!VisibilityListenerDef} callback
+   */
+  listenOnce(config, callback) {
+    const elId = config['selector'];
+    const res = resourcesFor(this.win_).get().find(el => {
```

Comment:
Still looking through feedback and have not looked into this in depth but doesn't [getResourceForElement](https://github.com/ampproject/amphtml/blob/master/src/service/resources-impl.js#L316) expect an `AmpElement`?